cli_tools/gce_vm_image_import/importer/inflater.go

```diff
@@ -40,6 +43,119 @@ type inflater interface {
 	traceLogs() []string
 }
 
+// inflaterFacade implements an inflater using other concrete implementations.
+type inflaterFacade struct {
+	mainInflater   inflater
+	shadowInflater inflater
+}
+
+// signals to control the verification towards shadow inflater
+const (
+	sigMainInflaterDone   = "main done"
+	sigMainInflaterErr    = "main err"
+	sigShadowInflaterDone = "shadow done"
+	sigShadowInflaterErr  = "shadow err"
+)
+
+func (facade inflaterFacade) inflate(ctx context.Context) (persistentDisk, error) {
+	inflaterChan := make(chan string)
+
+	// Launch main inflater.
+	var pd persistentDisk
+	var err error
+	go func() {
+		pd, err = facade.mainInflater.inflate(ctx)
+		if err != nil {
+			inflaterChan <- sigMainInflaterErr
+		} else {
+			inflaterChan <- sigMainInflaterDone
+		}
+	}()
+
+	if facade.shadowInflater == nil {
+		pd.matchResult = "NA"
+		// Wait for main inflater to finish
+		<-inflaterChan
+		return pd, err
+	}
+
+	// Launch shadow inflater.
+	var shadowPd persistentDisk
+	var shadowErr error
+	go func() {
+		shadowPd, shadowErr = facade.shadowInflater.inflate(ctx)
+		if shadowErr != nil {
+			inflaterChan <- sigShadowInflaterErr
+		} else {
+			inflaterChan <- sigShadowInflaterDone
+		}
+	}()
+
+	// Return early if main inflater finished first.
+	result := <-inflaterChan
+	if result == sigMainInflaterDone || result == sigMainInflaterErr {
+		if result == sigMainInflaterDone {
```

Comment:
It's *possible* that the shadow inflater could be blocking here, waiting to write its result to the channel, right? I'm not sure it's a big deal, since I think you're assuming that the API call will be much faster than the daisy worklfow.

I guess they'd both have to finish simultaneously to allow that to happen

Reply:
Both shadow and main inflater send a signal to inflaterChan. So nothing will be blocked:
1. If the main inflater finished first (corner case), the main thread will return without waiting for shadow.
2. If the shadow inflater finsihed first, the main thread will wait for main inflater.
Nobody will wait for shadow inflater.